cmd/influx_inspect/cardinality/cardinality.go

```diff
@@ -0,0 +1,202 @@
+	"github.com/influxdata/influxdb/tsdb/engine/tsm1"
+)
+
+// Command represents the program execution for "influxd cardinality".
+type Command struct {
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+	Stdout io.Writer
+
+	dbPath     string
+	shardPaths map[uint64]string
+	exact      bool
+	// How many goroutines to dedicate to calculating cardinality.
+	concurrency int
+}
+
+// NewCommand returns a new instance of Command with default setting applied.
+func NewCommand() *Command {
+	return &Command{
+		Stderr:      os.Stderr,
+		Stdout:      os.Stdout,
+		shardPaths:  map[uint64]string{},
+		concurrency: runtime.GOMAXPROCS(0),
+	}
+}
+
+// Run executes the command.
+func (cmd *Command) Run(args ...string) (err error) {
+	fs := flag.NewFlagSet("cardinality", flag.ExitOnError)
+	fs.StringVar(&cmd.dbPath, "db-path", "", "Path to databaseRetentionPolicies. Required.")
+	fs.IntVar(&cmd.concurrency, "c", runtime.GOMAXPROCS(0), "Set worker concurrency. Defaults to GOMAXPROCS setting.")
+	fs.BoolVar(&cmd.exact, "exact", false, "Report exact counts")
+
+	fs.SetOutput(cmd.Stdout)
+	if err := fs.Parse(args); err != nil {
+		return err
+	}
+
+	if cmd.dbPath == "" {
+		return errors.New("path to databaseRetentionPolicies must be provided")
+	}
+
+	dbMap := make(databaseRetentionPolicies)
+
+	// Walk databaseRetentionPolicies directory to get shards.
+	if err := filepath.Walk(cmd.dbPath, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if !info.IsDir() {
+			return nil
+		}
+
+		// TODO(edd): this would be a problem if the retention policy was named
+		// "index".
+		if info.Name() == tsdb.SeriesFileDirectory || info.Name() == "index" {
+			return filepath.SkipDir
+		}
+
+		id, err := strconv.Atoi(info.Name())
+		if err != nil {
+			return nil
+		}
+		cmd.shardPaths[uint64(id)] = path
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if len(cmd.shardPaths) == 0 {
+		_, err := fmt.Fprintf(cmd.Stderr, "No shards under %s\n", cmd.dbPath)
+		return err
+	}
+
+	estTitle := " (estimated)"
+	newCounterFn := report.NewHLLCounter
+	if cmd.exact {
+		newCounterFn = report.NewExactCounter
+		estTitle = ""
+	}
+
+	for _, p := range cmd.shardPaths {
+		err := reporthelper.WalkShardDirs(p, func(db, rp, id, path string) error {
+			file, err := os.OpenFile(path, os.O_RDONLY, 0600)
+			if err != nil {
+				fmt.Fprintf(cmd.Stderr, "error: %s: %v. Skipping.\n", path, err)
+				return nil
+			}
+
+			reader, err := tsm1.NewTSMReader(file)
+			if err != nil {
+				fmt.Fprintf(cmd.Stderr, "error: %s: %v. Skipping.\n", file.Name(), err)
+				_ = file.Close()
+				return nil
+			}
+			seriesCount := reader.KeyCount()
+			for i := 0; i < seriesCount; i++ {
+				key, _ := reader.KeyAt(i)
+				seriesKey, field, _ := bytes.Cut(key, []byte("#!~#"))
+				measurement, _ := models.ParseKey(seriesKey)
+				fs := initFieldsAndSeries(dbMap, db, rp, measurement, newCounterFn)
+				fs.series.Add(key)
```

Comment:
`key` here already includes the field - so is it already calculating in Cloud2 mode? If not, can we just do that instead of trying to do `(series keys without fields total) * (fields total)` ?

Reply:
I think your conclusion is correct, but I like printing the old and new series counts....

Reply:
We have to take the parsing cost to get the measurement, no matter what, so the work done is just the multiply, and the extra printing.

Reply:
It's much more accurate to not do the multiply though - with the multiply approach, if there is one series key with many fields, and many series keys with only one field, we would massively overestimate unless we actually count the live combinations.

Reply:
Ah, true!